AP Float Tutorial (#3)
tiwaria1/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ap_float/README.md

| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
| What you will learn               | Including and using the `ap_float` type <br> Using `ap_float` type to trade off mathematical accuracy for lesser resource utilization <br> Using various `ap_float` rounding modes and their effect on accuracy and resource utilization <br> Using the `ap_float `math functions for better quality of results
| Time to complete                  | 1 hour

## Purpose

This FPGA tutorial shows how to use the `ap_float` type with some simple examples and recommended best practices.

This data-type can be used in place of native floating point types to generate area efficient and optimized designs for the FPGA. For example, operations which do not utilize all of the bits the native types or desings which do not require all of the range and precision of native types are good candidates for replacement with the `ap_float` type.

This tutorial will present the following:

> 1. How to include the `ap_float` type and an overview of common ap_float use cases.

Reviewer: please replace all instances of _ap_float_ with _\`ap_float\`_.

2. A Polynomial Sine Approximation example which illustrates how to trade off mathematical accuracy for lesser FPGA resource utilization.
3. Rounding Mode and native type to ap_float type conversion examples which describe various `ap_float` rounding modes and their effect on accuracy and FPGA resource utilization.
4. A Quadratic Equation Solver example which show cases explicit `ap_float` math functions and how they can be used to replace mathematical operators like `*, /, +` and `-` for better quality of results.

> ### Simple Code Example

Reviewer: this (and all other subsections) should be promoted up a level.

An `ap_float` number can be defined as follows:

```cpp
ihc::ap_float<EW, MW> a;
```

> Here EW specifies the exponent width and MW specifies the mantissa width of the number. Optionally, another template parameter can be specified to set the rounding mode. For more details please refer to the section titled `Variable-Precision Integer and Floating-Point Support` in the Intel® oneAPI DPC++ FPGA Optimization Guide.

Reviewer: I think `EW` and `MW` should be monospaced

> Here EW specifies the exponent width and MW specifies the mantissa width of the number. Optionally, another template parameter can be specified to set the rounding mode. For more details please refer to the section titled `Variable-Precision Integer and Floating-Point Support` in the Intel® oneAPI DPC++ FPGA Optimization Guide.

Reviewer: I think that we should explicitly describe floating points, either here or in the other documentation, and include a picture like this: 
![image](https://user-images.githubusercontent.com/86071037/137539212-56038bd8-7f4b-4801-8921-621c611fed73.png)

To use this type in your code, you must include the following header:

```cpp
#include <sycl/ext/intel/ac_types/ap_float.hpp>
```

To use `ap_float` math functions, you must include the following header:

```cpp
#include <sycl/ext/intel/ac_types/ap_float_math.hpp>
```

> Additionally, you must use the flag `-qactypes` in order to ensure that the headers are correctly included and that the compiler links against the necessary libraries for emulation support.

Reviewer: please tell the user where the `-qactypes` or `/Qactypes` flag appears.

Author: Added a note, however I might have the -fintelfpga flag absorb -qactypes before this tutorial is released.

You can easily convert your existing designs that use native floating-point types to use `ap_float`: simply switch the original type. For math functions, `ap_float` has the "ihc_" prefix, you can simply switch your math functions accordingly, e.g. `sin(x)` should be changed to `ihc_sin(x)` for `ap_float`.

After the migration, you can use the area report to examine the area improvement of your design. In general, the line structure of the area report does not change. For example, instead of seeing a `X bit floating-point multiply` on the old design, the source line for the changed design would show `fpga.vpfp.mul`. 

You should confirm that the area used for the operation has indeed decreased from a Quartus compile. You should also make sure that the result of your design still meets your accuracy expectations through simulation.

> ### Overview of common use cases for ap_float

Reviewer: is it worth it to mention rounding modes here too? that's another use case for `ap_float`.

> ### Overview of common use cases for ap_float

Reviewer: some of your subtitles are Fully Capitalized For All Words, and Some only capitalize the first word.

Since the subtitles in this tutorial go up to 4 levels down, I think you should go with having all words capitalized because it makes it easier to tell a subtitle from text.

> You should consider migrating to `ap_float` types when you have precision requirements that differ from native float and double types, including both the range (number of exponent bits) and precision (number of mantissa bits) metrics. 

Reviewer: when talking about types, I think they should be in monospace font. So talking about native C++ `double` types should be monospace, but talking generically about double-precision floating point types should not

Double precision operations cannot be placed into a single hardened DSP block like single-precision operations, so double precision operations 

> are significantly more area intensive and uses more hardware resource. Moreover, `float` only has 23 bits of mantissa but `double` has 52, this could

Reviewer: ```suggestion
are significantly more area intensive and use more hardware resources. Moreover, `float` only has 23 bits of mantissa but `double` has 52, this could
```

be an overkill for applications that only seek a sweet spot in between.

Additionally, the built in subnormal support with native `double` type is area intensive and being able to turn subnormal support off can be great for reducing area utilization if the application does not consider very small subnormal numbers.

### Trading off mathematical accuracy for better resource utilization

> The kernels `ApproximateSineWithDouble` and `ApproximateSineWithDouble` implement a simple polynomial approximation of the sine function. 

Reviewer: ```suggestion
The kernels `ApproximateSineWithDouble` and `ApproximateSineWithAPFloat` implement a simple polynomial approximation of the sine function. 
```

The former uses `double` type to do so and the latter uses an `ap_float<11,44, Rnd>`. The `Rnd` rounding mode rounds towards zero. These two kernels will illustrate how to trade off accuracy for lesser FPGA resource utilization.

See the section `Examining the Reports` to go over the differences in resource utilization between these kernels. See the section `Example of Output` to see the difference in accuracy of results produced by these kernels.

> Note how the kernel function has been written once and invoked with different types to run two versions: one with `double` type and another with the `ap_float` type.

Reviewer: ```suggestion
Note how the kernel function within `RunSineApproximationKernel()` has been written once and the individual kernels are only differentiated by their input/output datatypes: `ApproximateSineWithDouble uses `double` datatypes and `ApproximateSineWithAPFLoat uses `ap_float` datatypes.
```

> ```cpp

Reviewer: I don't really like the rendered appearance of code indented within the code block. I would prefer that it be left-justified.

```
    indented code
    looks like this
```

compared to

```
un-indented code block
looks like this
```

Author: My code block had some space at the start, I removed that and now I think it looks much better.

  // Approximate with native double type
  RunSineApproximationKernel<double, ApproximateSineWithDouble>(q, input,
                                                                double_result);
  ...
  constexpr auto Rnd = ihc::fp_config::FP_Round::RZERO;
  using ap_float_double = ihc::ap_float<11, 44, Rnd>;

  RunSineApproximationKernel<ap_float_double, ApproximateSineWithAPFloat>(
      q, ap_float_input, ap_float_result);
```

> This is because `ap_float` is designed to fully blend in with native C++ types for syntax and semantic.

Reviewer: ```suggestion
This code re-use is because `ap_float` is designed to fully blend in with native C++ types for syntax and semantic.
```

### Conversion between native types and ap_float

> In normal floating-point FPGA applications, floating-point literals are represented as compile time constants and implemented as tie-offs (wires that directly connects to Gnd/Vcc) in RTL. This allows the construction of a constant to use no hardware resource in the FPGA flow.

Reviewer: ```suggestion
In normal floating-point FPGA applications, floating-point literals are represented as compile-time constants and implemented as tie-offs (wires that directly connects to Gnd/Vcc) in RTL. This allows the construction of a constant to use no hardware resources in the FPGA flow.
```

> However, `ap_float` types that have non-standard exponent and mantissa widths cannot be trivially converted from C++ native float or double literals. As a result, the construction of an `ap_float` type is not always free of hardware resource usage and sometimes logic for intermediate conversion is required to round the native floating-point constant to `ap_float`.

Reviewer: ```suggestion
However, `ap_float` types that have non-standard exponent and mantissa widths cannot be trivially converted from C++ native `float` or `double` literals. As a result, the construction of an `ap_float` type may sometimes require FPGA logic resources to round the native floating-point constant to the specified `ap_float`. This is called 'intermediate conversion'.
```

> It is important to understand when the intermediate conversions can occur. Conversion not only happens when you are explicitly casting numbers, it can also happen when you perform arithmetic operations on `ap_float` types with different precisions, since the operation needs to unify the types of the operands by promoting the less "dominant" types (types that have lower representable range). This is demonstrated by the code in kernel `ConversionKernelA`.

Reviewer: ```suggestion
It is important to understand when the intermediate conversions can occur. Conversion does not only happen when you are explicitly casting numbers: it can also happen when you perform arithmetic operations using `ap_float` types with different precisions. Intermediate conversions are necessary because the operation needs to unify the types of the operands by promoting the less "dominant" types (types that have lower representable range). This is demonstrated by the kernel code in the function `TestConversionKernelA`.
```

Reviewer: I think it's better to draw the reader's attention to the function where the kernel is defined since the token `ConversionKernelA` is defined at the top of your code, and you actually want the user to look at the code in `TestConversionKernelA()`.

#### Converting native numbers to ap_float numbers with minimal FPGA hardware resources

There are a few ways to generate compile-time `ap_float` constants that do not require any hardware implementation:
 
  1. Initializing `ap_float<8,23>` from `float` or `ap_float<11,52>` from `double` is just a direct bitwise copy (wires in RTL), so if the input `float`/`double` is a compile-time constant, the constructed `ap_float` is also a compile-time constant. You may want to extend these two types instead of the native float and double type if you want to use `ap_float` specific floating-point arithmetic controls (for example, the explicit binary operation presented in the section titled `ap_float_explicit_arithmetic`).
 

>   2. Converting from a constant to another `ap_float` that has rounding mode `FP_Round::ZERO` also results in a compile time constant. This rounding mode is also respected in a binary operation when promotion rounding is required. This is demonstrated by the code in `ConversionKernelB`.

Reviewer 2: asdf

>   2. Converting from a constant to another `ap_float` that has rounding mode `FP_Round::ZERO` also results in a compile time constant. This rounding mode is also respected in a binary operation when promotion rounding is required. This is demonstrated by the code in `ConversionKernelB`.

Reviewer: ```suggestion
  2. Converting from a constant to another `ap_float` that has rounding mode `FP_Round::ZERO` also results in a compile time constant. This rounding mode is also respected in a binary operation when promotion rounding is required. This is demonstrated by the kernel code in the function `TestConversionKernelB()`.
```

The kernel code in this tutorial contains comments that describe which operations result in generation of explicit cast operations and which do not.

Note:
  1. When assigning the result of the `convert_to` function to another `ap_float`, if the left hand side of the assignment has different exponent or mantissa widths than the ones specified in the `convert_to` function on the right hand side, another conversion can occur. 
 
  2. If your code performs computations on constant/literal native floating-point values, the compiler can sometimes combine them at compile time and save area. This is a compiler optimization technique called 'constant folding' or 'constant propagation'. Please note that this optimization does not work for `ap_float` even when the operands are constant. You should compute your constant arithmetics in native types or pre-compute them by hand.

> ### Using explicit ap_float math functions in place of mathematical operators

Reviewer: Promote up a level to ##

> #### Introduction

Reviewer: no need for this heading

In C++ applications, the basic binary operations have little expressiveness. On the contrary, FPGAs implement these operations using configurable logic, so you can improve your design's performance by fine-tuning the floating-point operations since they are usually area and latency intensive.

> The kernel `SpecializedQuadraticEqnSolverKernel` demonstrates how to use the explicit versions of `ap_float` binary operators to perform floating-point arithmetic operations based on your need.

Reviewer: T
```suggestion
The kernel code in the function `TestSpecializedQuadraticEqnSolver()` demonstrates how to use the explicit versions of `ap_float` binary operators to perform floating-point arithmetic operations based on your need.
```

I won't fix these anymore, but please update them.

 
To address these use cases, `ap_float` provides an explicit version of binary operators using template functions. The explicit operators provide 3 more main features in addition to basic binary operators.
 
1. Allow inputs and outputs with different precisions in the multiplication.

2. Tweak the area and accuracy trade off of the binary operations.

    The binary operations have high accuracy by default and produce results that are 0.5 ULP off from the most correct result. Users can override the default to choose an implementation with less area but also less precision (1 ULP).

3. Turn on/off subnormal support in the binary operations.

    To save area, subnormal support in the binary operators default to auto, this means it would be off unless there is direct hardened DSP support for it. Users can turn it on when the computation is expected to produce values close to 0, with some additional area.

The general form of explicit operations are as follows:
 
For addition, subtraction and division:

> Syntax:

Reviewer: ```suggestion
<add a newline>
Syntax:
```

```
  ap_float<E, M>::add/sub/div<AccuracyOption, SubnormalOption>(op1, op2)
```
Usage:

> * Rounds op1 and op2 to the specified E (exponent) and M (mantissa) widths

Reviewer: ```suggestion
* Rounds `op1` and `op2` to the specified `E` (exponent) and `M` (mantissa) widths
```

* Implements the operation with the provided accuracy and subnormal options.
* Returns the result with type `ap_float`<E, M>
 
For multiplication:
Syntax:
```

>   ap_float<E, M>::mul<AccuracyOption, SubnormalOption>(op1, op2)

Reviewer: why is this not bundled in with `add/sub/div`?

Reviewer: nvm I see that `mul` behaves differently.

```
Usage:

> * Leaves op1 and op2 intact

Reviewer: ```suggestion
* Leaves `op1` and `op2` intact
```

* Implements the operation with the provided accuracy and subnormal options.

> * Returns the result with type `ap_float`<E, M>

Reviewer: ```suggestion
* Returns the result with type `ap_float`<E, M>
```
```suggestion
* Returns the result with type `ap_float<E, M>`
```

> AccuracyOption is optional and can be one of the ENUMs below:

Reviewer: ```suggestion
The accuracy setting is optional and can be one of the `enum`s below:
```

```
  ihc::fp_config::FP_Accuracy::HIGH
  ihc::fp_config::FP_Accuracy::LOW
```
 

> SubnormalOption is optional can be one of the ENUMs below:

Reviewer: ```suggestion
The subnormal setting is optional can be one of the `enum`s below:
```

```
  ihc::fp_config::FP_Subnormal::ON
  ihc::fp_config::FP_Subnormal::OFF
```

Note:

> * Both ENUMs need to be compile time constants.

Reviewer: ```suggestion
* Both `enum`s need to be compile time constants.
```

> * You must specify AccuracyOption if you want to specify the SubnormalOption.

Reviewer: ```suggestion
* You must specify an accuracy setting if you want to specify a subnormal setting.
```

     ```
     nmake fpga_emu
     ```
   * Generate the optimization report: 
     ```
     nmake report
     ``` 
   * Compile for FPGA hardware (longer compile time, targets FPGA device):
     ```
     nmake fpga
     ``` 

*Note:* The Intel® PAC with Intel Arria® 10 GX FPGA and Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) do not support Windows*. Compiling to FPGA hardware on Windows* requires a third-party or custom Board Support Package (BSP) with Windows* support.
 
### In Third-Party Integrated Development Environments (IDEs)

You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*).
For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide)

## Examining the Reports

Locate the pair of `report.html` files in either:

* **Report-only compile**:  `ap_float_report.prj`
* **FPGA hardware compile**: `ap_float.prj`

### Examining the area reports for the Sine Approximation kernels

Navigate to the "Area Analysis-> Area Analysis of System" page. Click on the `Kernel System` line to expand it.

> Observe the difference in resource utilization of the kernels `ApproximateSineWithDouble` and `ApproximateSineWithAPFloat`.

Reviewer: Using kernel names here is OK because that's how they appear in the reports.

Expand the lines with the kernel names by clicking on them and expand the sub hierarchies to observe how the `add, mult` and `div`
operations use lesser resources for the `ApproximateSineWithAPFloat` kernel.

> You should observe an area reduction of up to 30% in resource utilization of the binary operations.

Reviewer: Is this area reduction regression tested?

> You should observe an area reduction of up to 30% in resource utilization of the binary operations.

Reviewer: ```suggestion
You should observe an area reduction in resource utilization of up to 30% for the binary operations.
```

perhaps include a before/after screenshot? I know my first instinct would be to compare the overall area in the summary report.

> TODO: Is simulation supported for customers yet?

Reviewer: good question...

  In the graph for the cluster under `ConversionKernelB`, you will find that it now only contains one "cast" node. This corresponds to the code:
  ```cpp
    x * num_accessor[0] + ...
  ```
  Although `x` and `num_accessor[0]` represent ap_floats constructed to use rounding mode `RZERO`, the result of this operation is cast to the higher precision `ap_float` using the default rounding mode `RNE` as the multiplication result is an operand for the next operation which uses higher precision.

  The other cast node is represented by a combination of `shift`, `select`, and `and` operations hence only one node labeled as "cast" is visible in the reports.
  
  The reduction in the number of cast nodes as compared to `ConversionKernelA` results in reduction of hardware resources used by `ConversionKernelB`.
  
  Observe the differences in the resource usage of these two kernels by navigating to the *Area Analysis of System* report (*Area Analysis* > *Area Analysis of System*) and looking at the entries under `Kernel System`.

2. Kernel: `ConversionKernelC`
  This kernel shows how to use the `convert_to` function and modify the rounding mode for a specific operation.

>   In the graph for the cluster under `Kernel_C`, you will find that it contains two "cast" nodes, corresponding to the conversions:

Reviewer: does it make sense to include screenshots? probably not since we change the report graphics so much.
